q2cli/cli.py

```diff
@@ -0,0 +1,104 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import click
+import qiime
+from qiime.sdk import PluginManager, SubprocessExecutor
+
+from . import __version__ as q2cli_version
+
+@click.group()
+@click.pass_context
+def cli(ctx):
+    ctx.obj['plugin_manager'] = PluginManager()
+
+@cli.command()
+def version():
+    click.echo('QIIME version: %s' % qiime.__version__)
+    click.echo('q2cli version: %s' % q2cli_version)
+
+@cli.command()
+@click.pass_context
+def plugins(ctx):
+    installed_plugins = ctx.obj['plugin_manager'].plugins.items()
+    if len(installed_plugins) == 0:
+        click.echo('No plugins are currently installed.\nYou can browse the '
+                   'official QIIME 2 plugins at: https://github.com/qiime2.')
+    else:
+        click.echo('Installed plugins:')
+        for name, plugin in installed_plugins:
+            click.echo(' %s %s (%s)' %
+                       (name, plugin.version, plugin.website))
+
+# should the QIIME types be hashable to support this without casting
+# to a string?
+_type_map = {'Int': click.INT, 'Str': click.STRING, 'Float': click.FLOAT}
+
+def build_commands(group):
+    plugin_manager = PluginManager()
+    for plugin_name, plugin in plugin_manager.plugins.items():
+        for workflow_name, workflow in plugin.workflows.items():
+            parameters = []
+            for ia_name, ia_type in workflow.signature.input_artifacts.items():
+                p = click.Option(['--%s' % ia_name],
+                                   required=True,
+                                   type=click.Path(exists=True, dir_okay=False),
+                                   help='Input %s' % str(ia_type))
+                parameters.append(p)
+            for ip_name, ip_type in workflow.signature.input_parameters.items():
+                p = click.Option(['--%s' % ip_name],
+                                   required=True,
+                                   type=_type_map[str(ip_type)],
+                                   help=str(ip_type))
+                parameters.append(p)
+            for oa_name, oa_type in workflow.signature.output_artifacts.items():
+                p = click.Option(['--%s' % oa_name],
+                                   required=True,
+                                   type=click.Path(exists=False, dir_okay=False),
+                                   help='Output %s' % str(oa_type))
+                parameters.append(p)
+
+            def create_callback(wf):
```

Comment:
is it necessary that this method is define here vs. at module scope? I'm not seeing the ties to local variables but might be missing it.